File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
    ArrayList<ListNode> nodeList = new ArrayList<>(); 
    ListNode currNode = head; 

    while(currNode != null){ 
        nodeList.add(currNode); 
        currNode = currNode.next;
    }

    if(n == nodeList.size()) {  // base case ig? 
        head = head.next; 
        return head; 
    }

    n = nodeList.size()-n; 
    ListNode curr = head; 

    int i = 0; 
    while(i != n){ 
        if(i == (n-1)){ 
            ListNode nNext = curr.next.next; 
            curr.next = nNext; 
        }
        i++; 
        curr = curr.next; 
    }

    return head; 
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        groupPrev = dummy 
        
        while(True): 
            kthNode = self.getKth(groupPrev, k)
            if kthNode: 
                groupNext = kthNode.next 

                prev, curr = kthNode.next, groupPrev.next 

                while curr != groupNext: 
                    next = curr.next 
                    curr.next = prev 
                    curr = next 
                    prev = curr

                temp = groupPrev.next 
                groupPrev.next = kthNode
                groupPrev = tmp 
            else: 
                break

        return dummy.next 

    def getKth(self, curr, k): 
        while curr and k > 0: 
            curr = curr.next 
            k -= 1 
        return curr 
